x-pack/platform/packages/shared/kbn-streamlang/src/schema/get_json_schema_from_streamlang_schema.ts

```diff
@@ -0,0 +1,501 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+// NOTE: Some of the workaround logic in this file (for fixBrokenSchemaReferencesAndEnforceStrictValidation) is similar
+// to that found in src/platform/packages/shared/kbn-workflows/spec/lib/get_json_schema_from_yaml_schema.ts
+// from Workflows. We may be able to align / share these utilities in the future once both projects aren't
+// in flux.
+import type { JsonSchema7Type } from 'zod-to-json-schema';
+import { zodToJsonSchema } from 'zod-to-json-schema';
+import type { z } from '@kbn/zod';
+import { ACTION_METADATA_MAP } from '../actions/action_metadata';
+
+/**
+ * JSON Schema scaffold produced from the Streamlang Zod schema. The converter
+ * always emits a root `$ref` that points into `definitions.StreamlangSchema`,
+ * so we capture that structure here to provide type safety when we later
+ * mutate the generated schema tree.
+ */
+type StreamlangJsonSchema = JsonSchema7Type & {
+  $ref: '#/definitions/StreamlangSchema';
+  $schema: 'http://json-schema.org/draft-07/schema#';
+  definitions: {
+    StreamlangSchema: JsonSchema7Type;
+  };
+};
+
+/**
+ * Convert the Streamlang Zod schema into JSON Schema and run our fixup pipeline
+ * so the result is consumable by Monaco YAML and other validation tooling.
+ *
+ * The raw output of `zod-to-json-schema` contains recursive references that
+ * Monaco cannot resolve, and it omits strict `additionalProperties` flags.
+ * This helper normalises those quirks before returning the schema artifact.
+ */
+export function getJsonSchemaFromStreamlangSchema(
```

Comment:
Quite a lot of ceremony in here because we don't have a discriminator on our step types (between `action` vs `where` steps), which trips up the JSON schema. Then some of the additional functionality is so that we get a nice display in the Monaco menu.